File: Java/src/main/java/UnionFind.java
public class UnionFind {
    private int[] pointer;
    private int[] size;
    private long total;

    public UnionFind(int size) {
        this.pointer = new int[size];
        this.size = new int[size];
        this.total = size;
        for (int i = 0; i < pointer.length; i++)
            pointer[i] = i;
    }

    public int find(int a) {
        while (pointer[a] != a) a = pointer[a];
        return a;
    }

    public void union(int a, int b) {
        var topA = find(a);
        var topB = find(b);
        if (topA == topB) return;

        this.total--;

        if (size[topA] > size[topB]) {
            pointer[topB] = pointer[topA];
            size[topA] += size[topB];
        } else {
            pointer[topA] = pointer[topB];
            size[topB] += size[topA];
        }
    }

    public boolean isConnected(int a, int b) {
        return find(a) == find(b);
    }

    public long getTotal() {
        return total;
    }
}

File: C++/src/UnionFind.cpp
#include <vector>

using namespace std;

class UnionFind {
private:
    vector<int> pointer;
    vector<int> size;
    long total;

public:
    explicit UnionFind(int size) {
        this->total = size;
        for (int i = 0; i < size; i++) {
            pointer.push_back(i);
            this->size.push_back(0);
        }
    }

    int find(int a) {
        while (pointer.at(a) != a) a = pointer.at(a);
        return a;
    }

    bool isConnected(int a, int b) {
        return find(a) == find(b);
    }

    void connect(int a, int b) {
        int topA = find(a);
        int topB = find(b);
        if (topA == topB) return;

        this->total--;

        if (size.at(topA) > size.at(topB)) {
            pointer.at(topB) = pointer.at(topA);
            size.at(topA) += size.at(topB);
        }
        else {
            pointer.at(topA) = pointer.at(topB);
            size.at(topB) += size.at(topA);
        }
    }

    long getTotal() {
        return total;
    }
};